ci2/ci/build.py

```diff
@@ -0,0 +1,672 @@
+        return [self.job.id]
+
+    @staticmethod
+    def from_json(pr, name, deps, json):
+        return RunImageStep(pr, name, deps,
+                            json['image'],
+                            json['script'],
+                            json.get('inputs'),
+                            json.get('outputs'))
+
+    def config(self):  # pylint: disable=no-self-use
+        return {}
+
+    async def build(self, batch, pr):
+        template = jinja2.Template(self.script, undefined=jinja2.StrictUndefined)
+        rendered_script = template.render(**self.input_config(pr))
+
+        log.info(f'step {self.name}, rendered script:\n{rendered_script}')
+
+        if self.inputs:
+            input_files = []
+            for i in self.inputs:
+                input_files.append((f'{BUCKET}/build/{batch.attributes["token"]}{i["from"]}', i["to"]))
+        else:
+            input_files = None
+
+        if self.outputs:
+            output_files = []
+            for o in self.outputs:
+                output_files.append((o["from"], f'{BUCKET}/build/{batch.attributes["token"]}{o["to"]}'))
+        else:
+            output_files = None
+
+        sa = None
+        if (self.outputs is not None) or (self.inputs is not None):
+            sa = 'ci2'
+
+        self.job = await batch.create_job(
+            self.image,
+            command=['bash', '-c', rendered_script],
+            attributes={'name': self.name},
+            input_files=input_files,
+            output_files=output_files,
+            copy_service_account_name=sa,
+            parent_ids=self.deps_parent_ids())
+
+    async def cleanup(self, batch, sink):
+        pass
+
+
+class CreateNamespaceStep(Step):
+    def __init__(self, pr, name, deps, namespace_name, admin_sa, public, secrets):
+        super().__init__(name, deps)
+        self.namespace_name = namespace_name
+        if admin_sa:
+            self.admin_sa = {
+                'name': admin_sa['name'],
+                # FIXME check
+                'namespace': expand_value_from(admin_sa['namespace'], self.input_config(pr))
+            }
+        else:
+            self.admin_sa = None
+        self.public = public
+        self.secrets = secrets
+        self.job = None
+        self._name = f'test-{pr.number}-{namespace_name}-{self.token}'
+
+    def parent_ids(self):
+        return [self.job.id]
+
+    @staticmethod
+    def from_json(pr, name, deps, json):
+        return CreateNamespaceStep(pr, name, deps,
+                                   json['namespaceName'],
+                                   json.get('adminServiceAccount'),
+                                   json.get('public', False),
+                                   json.get('secrets'))
+
+    def config(self):
+        conf = {'name': self._name}
+        if self.public:
+            conf['domain'] = f'{self._name}.internal.{DOMAIN}'
+        return conf
+
+    async def build(self, batch, pr):  # pylint: disable=unused-argument
```

Comment:
How do handle dependencies between steps where CI wants to run part of the pipeline, and batch needs to run part of the pipeline? I contemplated this path, it it seems like I duplicate the entire batch logic, dependency tracking, retries, cleanup, etc.  Maybe I can do it for creating namespaces (and force them to have no dependencies), but what about the deploy step?

The other alternative is for batch to support non-docker steps like creating namespaces and whatnot, then it can call the k8s API.

Honestly, yaml into kubectl seems intentionally equivalent to hitting the API.  That's all it does, so while I sympathize, this doesn't seem like a real difference.